File: brady/last_stone_weight.java
class Solution {
  
    // O(nlg(n)) time, O(n) space
    public int lastStoneWeight(int[] stones) {
        // Empty edge case
        if (stones.length == 0) return 0;

        // Store the stones in a min-heap
        List<Integer> maxHeap = new ArrayList<>(stones.length);
        for (int stone : stones) {
            insertNode(maxHeap, stone);
        }

        // Smash the stones iteratively
        while (maxHeap.size() > 1) {
            insertNode(maxHeap, Math.abs(extractNode(maxHeap)-extractNode(maxHeap)));
        }

        return maxHeap.get(0);
    }
  
    // Parent index: (i-1)/2
    // Left child index: 2i+1
    // Right child index: 2i+2

    private void insertNode(List<Integer> maxHeap, int node) {
        maxHeap.add(node);
        for (int i = maxHeap.size()-1; i > 0; i = (i-1)/2) {
            if (maxHeap.get((i-1)/2) >= maxHeap.get(i)) return;
            swap(maxHeap, i, (i-1)/2);
        }
    }
  
    private int extractNode(List<Integer> maxHeap) {
        swap(maxHeap, 0, maxHeap.size()-1);
        int extractedNode = maxHeap.remove(maxHeap.size()-1);

        // Heapify
        for (int i = 0; (2*i)+1 < maxHeap.size();) {
            // Left child is in the for loop condition so it must exist
            int leftVal = maxHeap.get((2*i)+1);
            int rightVal = (2*i)+2 < maxHeap.size() ? maxHeap.get((2*i)+2) : Integer.MIN_VALUE;
            if (Math.max(leftVal, rightVal) > maxHeap.get(i)) {
                if (leftVal > rightVal) {
                    swap(maxHeap, i, (2*i)+1);
                    i = (2*i)+1;
                } else {
                    swap(maxHeap, i, (2*i)+2);
                    i = (2*i)+2;
                }
            } else return extractedNode;
        }

        return extractedNode;
    }

    private void swap(List<Integer> maxHeap, int i1, int i2) {
        int temp = maxHeap.get(i1);
        maxHeap.set(i1, maxHeap.get(i2));
        maxHeap.set(i2, temp);
    }
}

File: cole/day18_last-stone-weight/last-stone-weight.c
void heapify(int *arr, int n, int i)
{
    int largest = i, left = 2 * i + 1, right = 2 * i + 2;
    if (left < n && arr[left] > arr[largest])
        largest = left;
    if (right < n && arr[right] > arr[largest])
        largest = right;
    if (largest != i)
    {
        int temp = arr[i];
        arr[i] = arr[largest];
        arr[largest] = temp;
        heapify(arr, n, largest);
    }
}

void buildheap(int *arr, int n)
{
    int i = n / 2 - 1;
    while (i >= 0)
    {
        heapify(arr, n, i);
        i--;
    }
}

int twosort(int *ab)
{
    int *b = ab + 1;
    if (*ab < *b)
    {
        int c = *b;
        *b = *ab;
        *ab = c;
    }
    return *ab;
}

#define ABS(x) (x < 0 ? (-x) : x)

int lastStoneWeight(int *stones, int stonesSize)
{
    for (int i = 0; i < stonesSize - 2; ++i)
    {
        buildheap(stones + i, stonesSize - i);
        stones[i + 1] = ABS(stones[i] - twosort(stones + i + 1));
    }
    if (stonesSize > 1)
    {
        twosort(stones + stonesSize - 2);
        stones[stonesSize - 1] = stones[stonesSize - 2] - stones[stonesSize - 1];
    }
    return stones[stonesSize - 1];
}
